File: src/models/vision/selection_state.py
"""Selection state enumeration for visual elements.

This module defines the SelectionState enum for representing the
selection/activation state of visual elements.
"""

from enum import Enum


class SelectionState(str, Enum):
    """Selection/activation state of a visual element.

    This state is used to identify currently active menu items,
    selected tabs, or disabled controls.

    Values:
        SELECTED: Currently selected/highlighted element
        NORMAL: Normal unselected state
        DISABLED: Disabled state (grayed out, not interactive)
    """

    SELECTED = "selected"
    NORMAL = "normal"
    DISABLED = "disabled"

    @classmethod
    def from_string(cls, value: str) -> 'SelectionState':
        """Create SelectionState from string with fuzzy matching.

        Args:
            value: String value to convert

        Returns:
            SelectionState enum instance

        Examples:
            >>> SelectionState.from_string("selected")
            <SelectionState.SELECTED: 'selected'>
            >>> SelectionState.from_string("active")  # Fuzzy match
            <SelectionState.SELECTED: 'selected'>
            >>> SelectionState.from_string("gray")  # Fuzzy match
            <SelectionState.DISABLED: 'disabled'>
        """
        value_lower = value.lower().strip()

        # Try exact match first
        try:
            return cls(value_lower)
        except ValueError:
            pass

        # Fuzzy mapping for common alternatives
        selected_aliases = {'active', 'highlighted', 'highlight', 'checked'}
        disabled_aliases = {'gray', 'grayed', 'dimmed', 'inactive', 'hidden'}

        if value_lower in selected_aliases:
            return cls.SELECTED
        elif value_lower in disabled_aliases:
            return cls.DISABLED
        else:
            return cls.NORMAL

    @classmethod
    def values(cls) -> list[str]:
        """Get all enum values as a list of strings.

        Returns:
            List of valid SelectionState values
        """
        return [e.value for e in cls]

    @classmethod
    def is_valid(cls, value: str) -> bool:
        """Check if a string value is a valid SelectionState.

        Args:
            value: String value to validate

        Returns:
            True if value is a valid SelectionState
        """
        return value.lower() in cls.values()

    def is_interactive(self) -> bool:
        """Check if this state represents an interactive element.

        Returns:
            True if the element is interactive (not disabled)
        """
        return self != self.DISABLED

    def is_active(self) -> bool:
        """Check if this state represents an active/selected element.

        Returns:
            True if the element is selected
        """
        return self == self.SELECTED

File: src/UniClaw.Core/Domain/Models/Vision/SelectionState.cs
namespace UniClaw.Core.Domain.Models.Vision;

/// <summary>
/// 视觉元素的激活/选择状态
/// </summary>
public enum SelectionState
{
    /// <summary>当前选中/高亮状态</summary>
    Selected,

    /// <summary>普通未选中状态</summary>
    Normal,

    /// <summary>禁用状态（灰色，不可交互）</summary>
    Disabled
}

/// <summary>
/// SelectionState 扩展方法
/// </summary>
public static class SelectionStateExtensions
{
    // ── Selected aliases (5 values, Python + C# extension) ──
    private static readonly HashSet<string> SelectedAliases = new()
    {
        "selected",      // exact enum value
        "active",        // Python alias
        "checked",       // Python alias
        "highlight",     // Python alias (previously missing)
        "highlighted",   // Python alias (previously missing)
    };

    // ── Disabled aliases (6 values, Python + C# extension) ──
    private static readonly HashSet<string> DisabledAliases = new()
    {
        "disabled",      // C# extension (semantic: disabled → Disabled is intuitive)
        "inactive",      // Python alias
        "hidden",        // Python alias
        "gray",          // Python alias
        "grayed",        // Python alias
        "dimmed",        // Python alias
    };

    /// <summary>判断是否可交互（非禁用）</summary>
    public static bool IsInteractive(this SelectionState state) => state != SelectionState.Disabled;

    /// <summary>判断是否为选中状态</summary>
    public static bool IsActive(this SelectionState state) => state == SelectionState.Selected;

    /// <summary>
    /// 从字符串创建 SelectionState：精确别名 HashSet 匹配。
    /// DisabledAliases 先查（避免 "inactive" 误落入 "active" 子串匹配）；
    /// SelectedAliases 后查；未命中回落 Normal。
    /// 大小写不敏感。
    /// </summary>
    public static SelectionState FromString(string value)
    {
        var key = value.ToLowerInvariant().Trim();

        // Disabled 先查：防止 "inactive" 被误判为 Selected (含 "active" 子串)
        if (DisabledAliases.Contains(key))
            return SelectionState.Disabled;

        if (SelectedAliases.Contains(key))
            return SelectionState.Selected;

        return SelectionState.Normal;
    }

    /// <summary>
    /// 判断字符串能否被 FromString 成功解析为已知状态（非回落 Normal）。
    /// 含别名 + "normal" 精确值；未知值返回 false。
    /// </summary>
    public static bool IsValid(string value)
    {
        if (string.IsNullOrEmpty(value)) return false;
        var key = value.ToLowerInvariant().Trim();
        return SelectedAliases.Contains(key) || DisabledAliases.Contains(key) || key == "normal";
    }
}
